### PR TITLE
fix: Ensure shared references in rule configs are separated.

### LINKED ISSUE #12592: Schema validation modifies exported config object in .eslintrc.js
**Tell us about your environment**

* **ESLint Version:** 6.7.0
* **Node Version:** 12.13.1
* **npm Version:** 6.12.1

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
// .eslintrc.basic.js
module.exports = {
  plugins: [],

  extends: [
    'eslint:all'
  ],

  parser: 'espree',

  parserOptions: {
    ecmaVersion: 7,
    sourceType: 'module'
  },

  settings: {},

  env: {
    es6: true
  },

  rules: {
    'camelcase': ['warn', {properties: 'never'}]
  }
}

// .eslintrc.babel.js
let _basic = require('./.eslintrc.basic');

module.exports = {
  parser: 'babel-eslint',

  parserOptions: {
    sourceType: 'module'
  },

  plugins: [
    'babel'
  ],

  rules: {
    camelcase: 'off',
    'babel/camelcase': _basic.rules.camelcase
  }
};

// .eslintrc.js
module.exports: {
  extends: [
    './.eslintrc.basic.js',
    './.eslintrc.babel.js'
  ]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
N/A
```

```bash
N/A
```

**What did you expect to happen?**

I expected no errors.

**What actually happened? Please include the actual, raw output from ESLint.**

```
	Configuration for rule "babel/camelcase" is invalid:
	Value {"properties":"never","ignoreDestructuring":false,"ignoreImports":false} should NOT have additional properties.
```

This happens because of https://github.com/eslint/eslint/pull/12528 adding `ignoreImports` with default value `failse`, and because the defaults of `camelcase` are merged into the same object in `.eslintrc.basic.js:rules.camelcase`, which is also the same reference in `.eslintrc.babel.js:rules.@babel/camelcase`.

**Are you willing to submit a pull request to fix this bug?**

Sure.

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

In `eslint.config.js` files, it's possible for two rule configs to contain references to the same object. That object may be modified during validation, thus affecting all configs, and may create validation errors. This clones rule configs before validation to ensure that each rule is using unique references to manage its configs.

Cloning works using the `structuredClone` function, so all values except for functions will be cloned.

Fixes #12592

### CODE REVIEW

File: lib/config/flat-config-schema.js [hunk 1]
```diff
@@ -5,6 +5,16 @@
...
 
+//-----------------------------------------------------------------------------
+// Requirements
+//-----------------------------------------------------------------------------
+
+/*
+ * Note: This can be removed in ESLint v9 because structuredClone is available globally
+ * starting in Node.js v17.
+ */
+const { default: structuredClone } = require("@ungap/structured-clone");
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:16]: Looks like this causes problems for wdio, this is an error we're getting on CI browser test:

```
Error in "Linter with FlatConfigArray.verify().Rule Severity.rule should run as warning when set to 1 with a config array"
Error: Key "rules": Key "semi": structuredClone is not a function
    at http://localhost:40387/node_modules/@humanwhocodes/object-schema/src/object-schema.js:281:27
    at Array.reduce (<anonymous>)
    at _ObjectSchema.merge (http://localhost:40387/node_modules/@humanwhocodes/object-schema/src/object-schema.js:268:24)
    at http://localhost:40387/node_modules/@humanwhocodes/config-array/api.js:966:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (http://localhost:40387/node_modules/@humanwhocodes/config-array/api.js:965:39)
    at Linter._verifyWithFlatConfigArray (http://localhost:40387/build/eslint.js:2:61[45](https://github.com/eslint/eslint/actions/runs/6577334057/job/17871071555#step:5:46)5)
    at Linter.verify (http://localhost:40387/build/eslint.js:2:56041)
    at Context.call (http://localhost:40387/home/runner/work/eslint/eslint/tests/lib/linter/linter.js:12136:41)
```

I'm getting the same errors locally, but it works (at least for me locally) when switched to this form:

```suggestion
const structuredClone = require("@ungap/structured-clone").default;
```

File: tests/lib/config/flat-config-array.js [hunk 2]
```diff
@@ -1987,4 +1987,68 @@ describe("FlatConfigArray", () => {
...
 
     });
+
+    // https://github.com/eslint/eslint/issues/12592
+    describe("Shared references between rule configs", () => {
+
+        it("shared rule config should not cause a rule validation error", () => {
+
+            /* eslint-disable-next-line prefer-const -- need to allow mutation inside FlatConfigArray */
+            let ruleConfig = ["error", {}];
```
Comment (mdjermanovic) [inline tests/lib/config/flat-config-array.js:1997]:
```suggestion
            const ruleConfig = ["error", {}];
```

I think `let` isn't necessary here.

File: tests/lib/config/flat-config-array.js [hunk 3]
```diff
@@ -1987,4 +1987,68 @@ describe("FlatConfigArray", () => {
...
+            });
+
+        });
+
+
+        it("should throw rule validation error for camelcase", async () => {
+
+
+            /* eslint-disable-next-line prefer-const -- need to allow mutation inside FlatConfigArray */
+            let ruleConfig = ["error", {}];
```
Comment (mdjermanovic) [inline tests/lib/config/flat-config-array.js:2026]:
```suggestion
            const ruleConfig = ["error", {}];
```

File: tests/lib/config/flat-config-array.js [hunk 4]
```diff
@@ -1987,4 +1987,68 @@ describe("FlatConfigArray", () => {
...
+                    "default-case": ruleConfig,
+
+                    /* eslint-disable-next-line no-dupe-keys -- needed for testing */
+                    camelcase: [
+                        "error",
+                        {
+                            ignoreDestructuring: Date
+                        }
+
+                    ]
```
Comment (mdjermanovic) [inline tests/lib/config/flat-config-array.js:2040]:
```suggestion
                    "default-case": ruleConfig,

                    camelcase: [
                        "error",
                        {
                            ignoreDestructuring: Date
                        }

                    ]
```

I believe duplicate keys don't affect this testing, the previous one is just overwritten in the created object. Perhaps the configuration was intended to have two config objects? (though the test might be even better with one to make sure we're always cloning rule configs)

Comment (nzakas) [reply tests/lib/config/flat-config-array.js:2040]: I copied this from the tests that were originally submitted for fixing this bug. Adding a second config would make sense. :+1:

File: lib/config/flat-config-schema.js (hunk 1 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:16]: We can fix it to make things work, but seems like we should ping @christian-bromann to let him know that this is an issue. I can't imagine why destructuring wouldn't work.

File: lib/config/flat-config-schema.js [hunk 5]
```diff
@@ -119,7 +129,7 @@ function normalizeRuleOptions(ruleOptions) {
         : [ruleOptions];
 
     finalOptions[0] = ruleSeverities.get(finalOptions[0]);
-    return finalOptions;
+    return finalOptions.map(value => structuredClone(value));
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:132]:
```suggestion
    return structuredClone(finalOptions);
```

We could simplify it this way?

File: lib/config/flat-config-schema.js (hunk 1 above)
Comment (christian-bromann) [reply lib/config/flat-config-schema.js]: I can't remember what particular error I experience but this becomes problematic when transforming into ESM, let me take a look

File: package.json [hunk 6]
```diff
@@ -64,9 +64,10 @@
     "@eslint-community/regexpp": "^4.6.1",
     "@eslint/eslintrc": "^2.1.2",
     "@eslint/js": "8.51.0",
-    "@humanwhocodes/config-array": "^0.11.11",
+    "@humanwhocodes/config-array": "^0.11.12",
```
Comment (mdjermanovic) [inline package.json:67]:
```suggestion
    "@humanwhocodes/config-array": "^0.11.13",
```

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**